core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -94,19 +104,60 @@ object RequestChannel extends Logging {
     @volatile var recordNetworkThreadTimeCallback: Option[Long => Unit] = None
 
     val session = Session(context.principal, context.clientAddress)
+
     private val bodyAndSize: RequestAndSize = context.parseRequest(buffer)
 
     def header: RequestHeader = context.header
     def sizeOfBodyInBytes: Int = bodyAndSize.size
 
-    //most request types are parsed entirely into objects at this point. for those we can release the underlying buffer.
-    //some (like produce, or any time the schema contains fields of types BYTES or NULLABLE_BYTES) retain a reference
-    //to the buffer. for those requests we cannot release the buffer early, but only when request processing is done.
+    def buildResponse(abstractResponse: AbstractResponse,
+                      error: Errors): Send =
+      if (envelopeContext.isDefined) {
+        // Right now only the NOT_CONTROLLER error gets handled by the forwarding
+        // broker retry. Other errors should be fatal and propagated to the client.
+        val envelopeError = if (error.equals(Errors.NOT_CONTROLLER))
+          Errors.NOT_CONTROLLER
+        else
+          Errors.NONE
+
+        val envelopeResponse = new EnvelopeResponse(
+          abstractResponse.throttleTimeMs(),
+          abstractResponse.serializeBody(context.header.apiVersion),
+          envelopeError
+        )
+        envelopeContext.get.brokerContext.buildResponse(envelopeResponse)
+      } else
+        context.buildResponse(abstractResponse)
+
+    def responseString(response: AbstractResponse): Option[String] =
+      if (RequestChannel.isRequestLoggingEnabled) {
+        Some(if (envelopeContext.isDefined)
+          response.toString(envelopeContext.get.brokerContext.apiVersion)
+        else
+          response.toString(context.apiVersion))
+      } else
+        None
+
+    def headerForLogging(): RequestHeader =
+      if (envelopeContext.isDefined)
+        envelopeContext.get.brokerContext.header
+      else
+        context.header
+
+    // most request types are parsed entirely into objects at this point. for those we can release the underlying buffer.
```

Comment:
nit: can we move this back to where the request parsing logic is. Otherwise it becomes a bit hidden.